// sanity-plugin-dashboard-widget-netlify/dist/components/SiteItem/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IMAGE_PULL_INTERVAL = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _Links = _interopRequireDefault(require("./Links"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var IMAGE_PULL_INTERVAL = 10000;
exports.IMAGE_PULL_INTERVAL = IMAGE_PULL_INTERVAL;

var getImageUrl = (siteId, branchName) => {
  var baseUrl = "https://api.netlify.com/api/v1/badges/".concat(siteId, "/deploy-status");
  var time = new Date().getTime();
  var branch = "branch=".concat(branchName);
  return branchName ? "".concat(baseUrl, "?").concat(time, "&").concat(branch) : "".concat(baseUrl, "?").concat(time);
};

var useBadgeImage = (siteId, branchName) => {
  var _useState = (0, _react.useState)(() => getImageUrl(siteId, branchName)),
      _useState2 = _slicedToArray(_useState, 2),
      src = _useState2[0],
      setSrc = _useState2[1];

  var update = (0, _react.useCallback)(() => setSrc(getImageUrl(siteId, branchName)), [siteId]);
  (0, _react.useEffect)(() => {
    var interval = window.setInterval(update, IMAGE_PULL_INTERVAL);
    return () => window.clearInterval(interval);
  }, [update]);
  return [src, update];
};

var useDeploy = (site, onDeploy, updateBadge) => {
  var timeoutRef = (0, _react.useRef)(-1);
  (0, _react.useEffect)(() => () => window.clearTimeout(timeoutRef.current), []);
  return (0, _react.useCallback)(() => {
    onDeploy(site);
    timeoutRef.current = window.setTimeout(updateBadge, 1000);
  }, [site, onDeploy, updateBadge]);
};

var SiteItem = props => {
  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      hasBadgeError = _useState4[0],
      setHasBadgeError = _useState4[1];

  var site = props.site,
      onDeploy = props.onDeploy;
  var id = site.id,
      name = site.name,
      title = site.title,
      url = site.url,
      adminUrl = site.adminUrl,
      buildHookId = site.buildHookId,
      branch = site.branch;

  var _useBadgeImage = useBadgeImage(id, branch),
      _useBadgeImage2 = _slicedToArray(_useBadgeImage, 2),
      badge = _useBadgeImage2[0],
      updateBadge = _useBadgeImage2[1];

  var handleDeploy = useDeploy(site, onDeploy, updateBadge);

  var handleBadgeError = () => {
    setHasBadgeError(true);
  };

  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    as: "li"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    paddingY: 2,
    paddingX: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "h4"
  }, title || name, /*#__PURE__*/_react.default.createElement(_Links.default, {
    url: url,
    adminUrl: adminUrl
  })), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    justify: "flex-start"
  }, !hasBadgeError && /*#__PURE__*/_react.default.createElement("img", {
    src: badge,
    onError: handleBadgeError,
    alt: "Badge"
  }), hasBadgeError && /*#__PURE__*/_react.default.createElement(_ui.Card, {
    tone: "critical",
    radius: 2,
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    size: 0,
    muted: true
  }, "Failed to load badge"))))), buildHookId ? /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingY: 2,
    paddingX: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    mode: "ghost",
    onClick: handleDeploy,
    text: "Deploy"
  })) : null);
};

var _default = SiteItem;
exports.default = _default;
