// sanity/lib/_chunks/execScript-8614ec34.esm.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
import { spawn } from 'child_process';
import fs from 'fs/promises';
import path from 'path';
import yargs from 'yargs/yargs';
import { hideBin } from 'yargs/helpers';
import readPkgUp from 'read-pkg-up';
function parseCliFlags(args) {
  return yargs(hideBin(args.argv || process.argv)).command("exec [script]", "executes given script", cmd => cmd.positional("script", {
    type: "string",
    demandOption: true
  }).option("with-user-token", {
    type: "boolean",
    default: false
  }).option("mock-browser-env", {
    type: "boolean",
    default: false
  })).argv;
}
const execScript = async function execScript2(args, context) {
  var _a;
  const {
    withUserToken,
    mockBrowserEnv,
    script
  } = await parseCliFlags(args);
  const {
    workDir
  } = context;
  const scriptPath = path.resolve(script || "");
  if (!script) {
    throw new Error("SCRIPT must be provided. `sanity exec <script>`");
  }
  if (!(await fs.stat(scriptPath).catch(() => false))) {
    throw new Error("".concat(scriptPath, " does not exist"));
  }
  const sanityPkgPath = (_a = await readPkgUp({
    cwd: __dirname
  })) == null ? void 0 : _a.path;
  if (!sanityPkgPath) {
    throw new Error("Unable to resolve `sanity` module root");
  }
  const sanityDir = path.dirname(sanityPkgPath);
  const threadsDir = path.join(sanityDir, "lib", "_internal", "cli", "threads");
  const esbuildPath = path.join(threadsDir, "esbuild.js");
  const browserEnvPath = path.join(threadsDir, "registerBrowserEnv.js");
  const configClientPath = path.join(threadsDir, "configClient.js");
  if (!(await fs.stat(esbuildPath).catch(() => false))) {
    throw new Error("`sanity` module build error: missing threads");
  }
  const baseArgs = mockBrowserEnv ? ["-r", browserEnvPath] : ["-r", esbuildPath];
  const tokenArgs = withUserToken ? ["-r", configClientPath] : [];
  const nodeArgs = [...baseArgs, ...tokenArgs, scriptPath, ...args.extraArguments];
  const proc = spawn(process.argv[0], nodeArgs, {
    stdio: "inherit",
    env: _objectSpread(_objectSpread({}, process.env), {}, {
      SANITY_BASE_PATH: workDir
    })
  });
  proc.on("close", process.exit);
};
export { execScript as default };
