// sanity/lib/_chunks/index-b950ba89.js
'use strict';

const _excluded = ["index", "pane", "paneKey"],
  _excluded2 = ["child", "component", "menuItems", "menuItemGroups", "title", "type"];
var _templateObject;
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactIs = require('react-is');
require('sanity');
var ui = require('@sanity/ui');
var desk = require('./desk-9ef5c1f2.js');
var icons = require('@sanity/icons');
require('sanity/router');
require('@sanity/types');
require('lodash/omit.js');
require('@sanity/util/paths');
require('lodash/isNumber.js');
require('lodash/isString.js');
require('react-rx');
var index = require('./index-4edba1d4.js');
var styled = require('styled-components');
require('rxjs');
require('rxjs/operators');
require('react-copy-to-clipboard');
require('lodash/partition.js');
require('lodash/uniqBy.js');
require('@sanity/uuid');
require('lodash/camelCase.js');
require('lodash/uniq.js');
require('lodash/kebabCase.js');
require('lodash/pickBy.js');
require('lodash/find.js');
require('lodash/uniqueId.js');
require('lodash/startCase.js');
require('nanoid');
require('@sanity/generate-help-url');
require('lodash/isEqual.js');
require('is-hotkey');
require('@sanity/schema/_internal');
require('lodash/upperFirst.js');
require('@rexxars/react-json-inspector');
require('hashlru');
require('lodash/negate.js');
require('framer-motion');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);
function UserComponentPaneHeader(props) {
  const {
    actionHandlers,
    index: index$1,
    menuItems,
    menuItemGroups,
    title
  } = props;
  const {
    features
  } = index.useDeskTool();
  if (!(menuItems == null ? void 0 : menuItems.length) && !title) {
    return null;
  }
  return /* @__PURE__ */jsxRuntime.jsx(desk.PaneHeader, {
    actions: /* @__PURE__ */jsxRuntime.jsx(desk.PaneHeaderActions, {
      menuItems,
      menuItemGroups,
      actionHandlers
    }),
    backButton: features.backButton && index$1 > 0 && /* @__PURE__ */jsxRuntime.jsx(ui.Button, {
      as: desk.BackLink,
      "data-as": "a",
      icon: icons.ArrowLeftIcon,
      mode: "bleed"
    }),
    title
  });
}
const Root = styled__default["default"](ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: relative;\n"])));
function UserComponentPaneContent(props) {
  const {
    children
  } = props;
  const {
    collapsed
  } = desk.usePane();
  return /* @__PURE__ */jsxRuntime.jsx(Root, {
    hidden: collapsed,
    height: "fill",
    overflow: "auto",
    children
  });
}
function UserComponentPane(props) {
  const {
      index,
      pane,
      paneKey
    } = props,
    restProps = _objectWithoutProperties(props, _excluded);
  const {
      child,
      component,
      menuItems,
      menuItemGroups,
      title = "",
      type: _unused
    } = pane,
    restPane = _objectWithoutProperties(pane, _excluded2);
  const [ref, setRef] = React.useState(null);
  return /* @__PURE__ */jsxRuntime.jsxs(desk.Pane, {
    id: paneKey,
    minWidth: 320,
    selected: restProps.isSelected,
    children: [/* @__PURE__ */jsxRuntime.jsx(UserComponentPaneHeader, {
      actionHandlers: ref == null ? void 0 : ref.actionHandlers,
      index,
      menuItems,
      menuItemGroups,
      title
    }), /* @__PURE__ */jsxRuntime.jsxs(UserComponentPaneContent, {
      children: [reactIs.isValidElementType(component) && React.createElement(component, _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, restProps), restPane), {
        ref: setRef
      }), {}, {
        child,
        paneKey
      })), React.isValidElement(component) && component]
    })]
  });
}
exports["default"] = UserComponentPane;
